### DISCUSSION TITLE
A SSR setup that just works (for starting up)

### DISCUSSION THREAD

**Discussion by jcalfee** [solidjs/solid-start#1068] [Q&A] [answered]
Under `solid-start dev` routes/api, is there any way to exclude vite SSR pre-build and run the code that runs during SSR as if it were running directly under node (without the vite errors)?  During SSR, I need to start off with things working then move code into SSR as it is possible.  All packages and local code can't comply up-front, that is not pratical.  Is there a foolproof way to get routes/api working or even createServerData$ without all the trouble of having vite trying to transpile all the server-side code?  

This is a necessary step to getting started, without that I'm pretty stuck when it comes to building apps due to mysterious errors with long standing history of issues without clear resolutions and, I have found no document that covers all the loose ends like [optimizeDeps exclude](https://vite.nuxtjs.org/misc/common-issues/#uncaught-syntaxerror-the-requested-module--does-not-provide-an-export-named-default) and other errors like `__vite_ssr_import_0__.[MyFuncion] is not a function` and `ReferenceError: require is not defined` in any number of 3rd party non-module packages.  So, I guess I need to setup a two process setup to use solidjs.

Is there any way to start off using SSR just to improve rendering and the data-flow and without worrying about bundling yet?

Thank you!

**Comment by ryansolid** [comment, marked as answer]
There is an `ssr.external` option in Vite that might help. Vite tends have their own automatic handler that isn't always desired. https://vitejs.dev/config/ssr-options.html.

In terms of transpiling, if there is JSX or TypeScript there is going to be transpilation. Some adapters (non-node) also need bundling. Vite tries to be smart but it doesn't always work.

**Reply by jcalfee** [reply]
Thank you for this help.  I recognize this is outside of solidjs, I think a lot of this stems from node creating new situations that do not support their original features (not really JavaScript anymore).  I'm bringing it up over there, maybe we could all benefit from a  comparability fork of node.

When I use `ssr.noExternal = true` the `TypeError: __vite_ssr_import_0__.Db is not a function` goes away.  Now I'm stuck on the dreaded `require` bug (that is why I'm over at nodejs making issues):
```
An unhandled error occured: ReferenceError: require is not defined
    at /home/james/dev/ipall/node_modules/.pnpm/resolve-from@5.0.0/node_modules/resolve-from/index.js:2:14
    at instantiateModule (file:///home/james/dev/ipall/node_modules/.pnpm/vite@4.4.9_@types+node@18.17.0_sass@1.63.6/node_modules/vite/dist/node/chunks/dep-df561101.js:55974:15)
Sourcemap for "/home/james/dev/ipall/node_modules/.pnpm/@scaleleap+pg-format@1.0.0/node_modules/@scaleleap/pg-format/lib/index.js" points to missing source files
```

**Reply by jcalfee** [reply]
I think I'm starting to figure it out here, the errors are changing as I include more packages in `ssr.external`.  I have a dynamic import that loops around making this pretty interesting.  You have put me back on track!  Thank you..

**Reply by jcalfee** [reply]
> In terms of transpiling, if there is JSX or TypeScript there is going to be transpilation. 

I really like jsdoc ;).  It was all solved with ssr.external but it seamed to be very shaky (like my confidence at this point).  I just stayed focused on that as being the solution and it worked out.  So, I'm up and running.  thanks..

**Comment by jcalfee** [comment]
I have been at this an hour and I'm still not sure if this is vite or solid start.  I can't recreate it in Vite.  I'll try again tomorrow.  Here is what I see though:

After adding my locally linked package "ddl" to ssr.external, vite stopped trying to "handle" the code (now it does not error and runs great under `solid-start dev`, however, Vite also stopped re-starting the dev server when I save a file under the `ddl` package.  So, the goal is to have this package still under the watch but handled also as a ssr external resource.

The comments below are an attempt to re-add the package to file watch using a negated glob pattern "!...".

```js
  ssr: {
    external: [
      'ddl',
    ]
  },
  server: {
    watch: {
      // ignored: ['!**/../ddl/**'] // https://vitejs.dev/config/server-options.html#server-watch
    }
  },
  optimizeDeps: {
    // exclude: ['ddl'] // https://vitejs.dev/config/server-options.html#server-watch
  },
```

**Reply by jcalfee** [reply]
Turns out watch.ignored will accept a value [path=>{console.log(path)}] .. I can see from the console output, that when the package is in ssr.external it does not appear in the console output in watch.ignored.  So, it is not a glob issue it is simply not there.  I'll need to create a minimal environment to report / ask the vite team about this.  I may need to build a minimal repo to report it to vitejs and I may just fall-back to good old nodemon for now..
